# Data Cleaning Notebook

## Introduction

In this notebook, we will focus on the initial steps of preparing our dataset for analysis. Proper data cleaning is essential to ensure that the models built later are accurate and reliable. Specifically, we will address the following tasks:

1. **Rename Columns and add new columns**: Standardize column names for clarity and consistency.
2. **Verify Data Types**: Ensure that each column's data type aligns with the expected format for analysis (e.g., numeric, categorical).
3. **Check Uniqueness**: Identify any duplicate records or features that may affect our analysis.
4. **Handle Missingness**: Detect missing values and decide how to handle them (e.g., removing or imputing).
5. **Next Steps**

### Project Goal

The ultimate goal of this project is to build a portfolio of stocks that can beat the S&P 500 index. Using machine learning models, we aim to select stocks with the highest probability of outperforming the market based on financial indicators and historical data. By ensuring our dataset is clean and well-prepared, we will lay a solid foundation for accurate model training and testing, driving us closer to this goal.


### Importing Libraries

In [1]:
import pandas as pd
import numpy as np

### Loading Data Sets

In [2]:
# Load all datasets from 2014 to 2018
data_2014 = pd.read_csv('data/raw_data/2014_Financial_Data.csv')
data_2015 = pd.read_csv('data/raw_data/2015_Financial_Data.csv')
data_2016 = pd.read_csv('data/raw_data/2016_Financial_Data.csv')
data_2017 = pd.read_csv('data/raw_data/2017_Financial_Data.csv')
data_2018 = pd.read_csv('data/raw_data/2018_Financial_Data.csv')
# Data 2018 is already loaded as data_2018

full_df = {'2014': data_2014,
           '2015': data_2015,
           '2016': data_2016,
           '2017': data_2017,
           '2018': data_2018
          }

# Print the head of each year
for year, df in full_df.items():
    print(f"Head of {year}:")
    print(df.head(), "\n")

Head of 2014:
  Unnamed: 0       Revenue  Revenue Growth  Cost of Revenue  Gross Profit  \
0         PG  7.440100e+10         -0.0713     3.903000e+10  3.537100e+10   
1       VIPS  3.734148e+09          1.1737     2.805625e+09  9.285226e+08   
2         KR  9.837500e+10          0.0182     7.813800e+10  2.023700e+10   
3        RAD  2.552641e+10          0.0053     1.820268e+10  7.323734e+09   
4        GIS  1.790960e+10          0.0076     1.153980e+10  6.369800e+09   

   R&D Expenses  SG&A Expense  Operating Expenses  Operating Income  \
0  0.000000e+00  2.146100e+10        2.146100e+10      1.391000e+10   
1  1.083303e+08  3.441414e+08        7.939267e+08      1.345959e+08   
2  0.000000e+00  1.519600e+10        1.751200e+10      2.725000e+09   
3  0.000000e+00  6.561162e+09        6.586482e+09      7.372520e+08   
4  0.000000e+00  3.474300e+09        3.412400e+09      2.957400e+09   

   Interest Expense  ...  Receivables growth  Inventory Growth  Asset Growth  \
0      7.090000e

In [3]:
for year, df in full_df.items():
    print(f"{year} shape: ", df.shape)

2014 shape:  (3808, 225)
2015 shape:  (4120, 225)
2016 shape:  (4797, 225)
2017 shape:  (4960, 225)
2018 shape:  (4392, 225)


Let's Inspect the column names

In [4]:
full_df['2014'].columns.tolist()

['Unnamed: 0',
 'Revenue',
 'Revenue Growth',
 'Cost of Revenue',
 'Gross Profit',
 'R&D Expenses',
 'SG&A Expense',
 'Operating Expenses',
 'Operating Income',
 'Interest Expense',
 'Earnings before Tax',
 'Income Tax Expense',
 'Net Income - Non-Controlling int',
 'Net Income - Discontinued ops',
 'Net Income',
 'Preferred Dividends',
 'Net Income Com',
 'EPS',
 'EPS Diluted',
 'Weighted Average Shs Out',
 'Weighted Average Shs Out (Dil)',
 'Dividend per Share',
 'Gross Margin',
 'EBITDA Margin',
 'EBIT Margin',
 'Profit Margin',
 'Free Cash Flow margin',
 'EBITDA',
 'EBIT',
 'Consolidated Income',
 'Earnings Before Tax Margin',
 'Net Profit Margin',
 'Cash and cash equivalents',
 'Short-term investments',
 'Cash and short-term investments',
 'Receivables',
 'Inventories',
 'Total current assets',
 'Property, Plant & Equipment Net',
 'Goodwill and Intangible Assets',
 'Long-term investments',
 'Tax assets',
 'Total non-current assets',
 'Total assets',
 'Payables',
 'Short-term debt'

## 1. Rename Columns and Add New Columns

In this step, we will rename columns to standardize the naming conventions across all datasets. Consistent column names help simplify further analysis and ensure that operations can be applied uniformly to all datasets.

Additionally, we will add a new column, 'Year', to each dataset to clearly indicate the year each observation belongs to. This will help during data analysis and when merging datasets.

##### Let's add the year column to each dataset

In [5]:
for year, df in full_df.items():
    df['Year'] = int(year)
    

In [6]:
full_df['2014'].head(1)

Unnamed: 0       Revenue  Revenue Growth  Cost of Revenue  Gross Profit  \
0         PG  7.440100e+10         -0.0713     3.903000e+10  3.537100e+10   

   R&D Expenses  SG&A Expense  Operating Expenses  Operating Income  \
0           0.0  2.146100e+10        2.146100e+10      1.391000e+10   

   Interest Expense  ...  Inventory Growth  Asset Growth  \
0       709000000.0  ...           -0.0217        0.0359   

   Book Value per Share Growth  Debt Growth  R&D Expense Growth  \
0                       0.0316       0.1228                 0.0   

   SG&A Expenses Growth              Sector  2015 PRICE VAR [%]  Class  Year  
0               -0.1746  Consumer Defensive           -9.323276      0  2014  

[1 rows x 226 columns]

In [7]:
full_df['2015'].head(1)

Unnamed: 0       Revenue  Revenue Growth  Cost of Revenue  Gross Profit  \
0         PG  7.074900e+10         -0.0491     3.705600e+10  3.369300e+10   

   R&D Expenses  SG&A Expense  Operating Expenses  Operating Income  \
0           0.0  2.061600e+10        2.264400e+10      1.104900e+10   

   Interest Expense  ...  Inventory Growth  Asset Growth  \
0       626000000.0  ...           -0.2634       -0.1024   

   Book Value per Share Growth  Debt Growth  R&D Expense Growth  \
0                      -0.1013      -0.1432                 0.0   

   SG&A Expenses Growth              Sector  2016 PRICE VAR [%]  Class  Year  
0               -0.0394  Consumer Defensive           10.809451      1  2015  

[1 rows x 226 columns]

##### It looks like it is working well, now let's rename the Unnamed: 0 column to Symbol.

In [8]:
# Rename the 'Unnamed: 0' column to 'Symbol' in each dataset
for year, df in full_df.items():
    if 'Unnamed: 0' in df.columns:
        df.rename(columns={'Unnamed: 0': 'Symbol'}, inplace=True)


In [9]:
full_df['2014'].head(1)

Symbol       Revenue  Revenue Growth  Cost of Revenue  Gross Profit  \
0     PG  7.440100e+10         -0.0713     3.903000e+10  3.537100e+10   

   R&D Expenses  SG&A Expense  Operating Expenses  Operating Income  \
0           0.0  2.146100e+10        2.146100e+10      1.391000e+10   

   Interest Expense  ...  Inventory Growth  Asset Growth  \
0       709000000.0  ...           -0.0217        0.0359   

   Book Value per Share Growth  Debt Growth  R&D Expense Growth  \
0                       0.0316       0.1228                 0.0   

   SG&A Expenses Growth              Sector  2015 PRICE VAR [%]  Class  Year  
0               -0.1746  Consumer Defensive           -9.323276      0  2014  

[1 rows x 226 columns]

In [10]:
full_df['2015'].head(1)

Symbol       Revenue  Revenue Growth  Cost of Revenue  Gross Profit  \
0     PG  7.074900e+10         -0.0491     3.705600e+10  3.369300e+10   

   R&D Expenses  SG&A Expense  Operating Expenses  Operating Income  \
0           0.0  2.061600e+10        2.264400e+10      1.104900e+10   

   Interest Expense  ...  Inventory Growth  Asset Growth  \
0       626000000.0  ...           -0.2634       -0.1024   

   Book Value per Share Growth  Debt Growth  R&D Expense Growth  \
0                      -0.1013      -0.1432                 0.0   

   SG&A Expenses Growth              Sector  2016 PRICE VAR [%]  Class  Year  
0               -0.0394  Consumer Defensive           10.809451      1  2015  

[1 rows x 226 columns]

##### Now let's rename the column 'YEAR' PRICE VAR [%] to PRICE VAR [%]

In [11]:
# Rename columns that have the format '<year> PRICE VAR [%]' to 'PRICE VAR [%]'
for year, df in full_df.items():
    for col in df.columns:
        if 'PRICE VAR [%]' in col:
            df.rename(columns={col: 'PRICE VAR [%]'}, inplace=True)


In [12]:
full_df['2014'].head(1)

Symbol       Revenue  Revenue Growth  Cost of Revenue  Gross Profit  \
0     PG  7.440100e+10         -0.0713     3.903000e+10  3.537100e+10   

   R&D Expenses  SG&A Expense  Operating Expenses  Operating Income  \
0           0.0  2.146100e+10        2.146100e+10      1.391000e+10   

   Interest Expense  ...  Inventory Growth  Asset Growth  \
0       709000000.0  ...           -0.0217        0.0359   

   Book Value per Share Growth  Debt Growth  R&D Expense Growth  \
0                       0.0316       0.1228                 0.0   

   SG&A Expenses Growth              Sector  PRICE VAR [%]  Class  Year  
0               -0.1746  Consumer Defensive      -9.323276      0  2014  

[1 rows x 226 columns]

In [13]:
full_df['2015'].head(1)

Symbol       Revenue  Revenue Growth  Cost of Revenue  Gross Profit  \
0     PG  7.074900e+10         -0.0491     3.705600e+10  3.369300e+10   

   R&D Expenses  SG&A Expense  Operating Expenses  Operating Income  \
0           0.0  2.061600e+10        2.264400e+10      1.104900e+10   

   Interest Expense  ...  Inventory Growth  Asset Growth  \
0       626000000.0  ...           -0.2634       -0.1024   

   Book Value per Share Growth  Debt Growth  R&D Expense Growth  \
0                      -0.1013      -0.1432                 0.0   

   SG&A Expenses Growth              Sector  PRICE VAR [%]  Class  Year  
0               -0.0394  Consumer Defensive      10.809451      1  2015  

[1 rows x 226 columns]

##### Everything looks good so far.

## 2. Verify Data Types

In this step, we will verify the data types for each column in the dataset. Ensuring that each column has the correct data type is crucial for further analysis and modeling. For example, numerical values should be correctly typed as integers or floats, and categorical columns should be properly recognized as strings or categories.

We will check the data types for each year's dataset to ensure consistency and correct any discrepancies if found.


In [14]:
# Create a dataframe to display data types of all datasets side by side
data_types_df = pd.DataFrame()

for year, df in full_df.items():
    data_types_df[year] = df.dtypes

# Modify pandas display option to show all rows
pd.set_option('display.max_rows', None)

# Display the data types for each dataset side by side
display(data_types_df)

2014     2015     2016  \
Symbol                                       object   object   object   
Revenue                                     float64  float64  float64   
Revenue Growth                              float64  float64  float64   
Cost of Revenue                             float64  float64  float64   
Gross Profit                                float64  float64  float64   
R&D Expenses                                float64  float64  float64   
SG&A Expense                                float64  float64  float64   
Operating Expenses                          float64  float64  float64   
Operating Income                            float64  float64  float64   
Interest Expense                            float64  float64  float64   
Earnings before Tax                         float64  float64  float64   
Income Tax Expense                          float64  float64  float64   
Net Income - Non-Controlling int            float64  float64  float64   
Net Income - Discontinued ops               float64  float64  float64   
Net Income                                  float64  float64  float64   
Preferred Dividends                         float64  float64  float64   
Net Income Com                              float64  float64  float64   
EPS                                         float64  float64  float64   
EPS Diluted                                 float64  float64  float64   
Weighted Average Shs Out                    float64  float64  float64   
Weighted Average Shs Out (Dil)              float64  float64  float64   
Dividend per Share                          float64  float64  float64   
Gross Margin                                float64  float64  float64   
EBITDA Margin                               float64  float64  float64   
EBIT Margin                                 float64  float64  float64   
Profit Margin                               float64  float64  float64   
Free Cash Flow margin                       float64  float64  float64   
EBITDA                                      float64  float64  float64   
EBIT                                        float64  float64  float64   
Consolidated Income                         float64  float64  float64   
Earnings Before Tax Margin                  float64  float64  float64   
Net Profit Margin                           float64  float64  float64   
Cash and cash equivalents                   float64  float64  float64   
Short-term investments                      float64  float64  float64   
Cash and short-term investments             float64  float64  float64   
Receivables                                 float64  float64  float64   
Inventories                                 float64  float64  float64   
Total current assets                        float64  float64  float64   
Property, Plant & Equipment Net             float64  float64  float64   
Goodwill and Intangible Assets              float64  float64  float64   
Long-term investments                       float64  float64  float64   
Tax assets                                  float64  float64  float64   
Total non-current assets                    float64  float64  float64   
Total assets                                float64  float64  float64   
Payables                                    float64  float64  float64   
Short-term debt                             float64  float64  float64   
Total current liabilities                   float64  float64  float64   
Long-term debt                              float64  float64  float64   
Total debt                                  float64  float64  float64   
Deferred revenue                            float64  float64  float64   
Tax Liabilities                             float64  float64  float64   
Deposit Liabilities                         float64  float64  float64   
Total non-current liabilities               float64  float64  float64   
Total liabilities                           float64  float64  float64   
Other comprehensive income      

In [15]:
pd.reset_option('display.max_rows')

##### All types are in place. Let's move on

## 3. Check Uniqueness

In this step, we will check for any duplicate entries in our datasets. Ensuring uniqueness in the data is critical, as duplicate records can skew the analysis and lead to inaccurate model predictions. We will check for duplicate rows in each year's dataset.


In [16]:
# Check for uniqueness (duplicate rows) in each dataset
for year, df in full_df.items():
    duplicate_count = df.duplicated().sum()
    print(f"Number of duplicate rows in {year}: {duplicate_count}")

Number of duplicate rows in 2014: 0
Number of duplicate rows in 2015: 0
Number of duplicate rows in 2016: 0
Number of duplicate rows in 2017: 0
Number of duplicate rows in 2018: 0


##### There are no duplicate rows, everything is looking good until now.

## 4. Handle Missingness

In this step, we will address missing data in our datasets. Missing values can lead to biased results and reduce the effectiveness of our machine learning models. It's crucial to identify and handle missing data appropriately to maintain the integrity of our analysis.

We will:

- **Identify Missing Values**: Determine which columns and rows contain missing data.
- **Assess the Extent of Missingness**: Evaluate how much data is missing to decide on the handling strategy.
- **Handle Missing Data**:
  - **Remove Missing Values**: Drop rows or columns with excessive missing data if necessary.
  - **Impute Missing Values**: Fill in missing data using statistical methods like mean, median, or more advanced techniques.


In [17]:
# Create a dataframe to display the percentage of missingness for each feature year by year
missingness_df = pd.DataFrame()

for year, df in full_df.items():
    missingness_df[year] = df.isnull().mean() * 100

# Rearrange the index to display features (columns) and the percentage of missingness year by year
missingness_df.index.name = 'Feature'

In [18]:
# Modify pandas display option to show all rows
pd.set_option('display.max_rows', None)
missingness_df

2014       2015       2016  \
Feature                                                                       
Symbol                                       0.000000   0.000000   0.000000   
Revenue                                      1.155462   1.626214  10.193871   
Revenue Growth                               6.197479   8.252427  13.529289   
Cost of Revenue                              1.943277   4.684466  13.341672   
Gross Profit                                 1.365546   1.577670  10.256410   
R&D Expenses                                 3.571429   6.432039  14.863456   
SG&A Expense                                 1.549370   4.126214  12.778820   
Operating Expenses                           1.654412   4.223301  12.924745   
Operating Income                             1.444328   1.019417   9.693558   
Interest Expense                             1.654412   4.247573  12.924745   
Earnings before Tax                          2.100840   2.500000  11.069418   
Income Tax Expense                           1.733193   3.665049  12.424432   
Net Income - Non-Controlling int             3.912815   6.966019  15.384615   
Net Income - Discontinued ops                3.912815   6.966019  15.384615   
Net Income                                   0.603992   4.004854  12.632896   
Preferred Dividends                          3.912815   6.966019  15.384615   
Net Income Com                               0.393908   2.888350  11.632270   
EPS                                          1.890756   2.330097  10.881801   
EPS Diluted                                  2.205882   2.184466  10.735877   
Weighted Average Shs Out                     6.197479   2.839806  10.986033   
Weighted Average Shs Out (Dil)               2.888655   6.067961  14.383990   
Dividend per Share                           3.912815   6.966019  15.384615   
Gross Margin                                 1.234244   1.771845  10.402335   
EBITDA Margin                                5.304622   8.276699  16.552012   
EBIT Margin                                  1.313025   4.223301  12.945591   
Profit Margin                                5.278361   8.276699  16.552012   
Free Cash Flow margin                        4.175420   7.111650  15.530540   
EBITDA                                       1.680672   4.805825  13.404211   
EBIT                                         0.105042   2.961165  11.757348   
Consolidated Income                          0.603992   4.029126  12.653742   
Earnings Before Tax Margin                   1.234244   1.723301  10.360642   
Net Profit Margin                            1.496849   4.635922  13.279133   
Cash and cash equivalents                    1.523109   1.577670  10.277257   
Short-term investments                      24.501050   5.752427  13.383365   
Cash and short-term investments             24.422269   4.247573  12.174276   
Receivables                                  1.549370   1.043689   9.714405   
Inventories                                  3.019958   4.563107  13.279133   
Total current assets                        22.872899   4.830097  12.820513   
Property, Plant & Equipment Net              2.310924   3.762136  12.570356   
Goodwill and Intangible Assets               0.919118   2.888350  11.694809   
Long-term investments                       23.949580   5.024272  13.154055   
Tax assets                                   3.492647   5.655340  14.133834   
Total non-current assets                    23.844538  24.975728  31.436314   
Total assets                                 0.866597   2.135922  11.006879   
Payables                                     2.363445   3.834951  12.632896   
Short-term debt                             23.581933   5.728155  13.570982   
Total current liabilities                   23.082983   4.854369  12.820513   
Long-term debt                              22.426471   1.456311   9.651866   
Total debt                                   2.048319   4.514563  12.987284   
Deferred revenue    

### Analysis of Missing Data

After evaluating the percentage of missing data across features from 2014 to 2018, we observe that most features have varying degrees of missingness. The features with missing data range from 0% to over 40%. A significant number of features have missing data below 20%, while some show considerably higher levels of missingness, especially in financial growth metrics.

### Imputation Strategy

For features with less than 30% missingness, we will apply **KNN (K-Nearest Neighbors) Imputation** to fill in the missing values. This method considers the nearest neighbors to impute the missing data, preserving the integrity of the dataset, we will use median given the nature of the data wich must contain multiple natural outliers.

Once the KNN imputation is applied, we will recheck the dataset to ensure the missingness has been appropriately handled and that the data is ready for further analysis.


In [19]:
from sklearn.impute import KNNImputer

# Initialize KNN Imputer with 5 nearest neighbors
imputer = KNNImputer(n_neighbors=15)

# Apply KNN imputation to each dataset for numerical columns with less than 20% missingness
for year, df in full_df.items():
    # Select numerical columns with less than 20% missing values
    numerical_cols = df.select_dtypes(include=['float64', 'int64']).columns
    cols_to_impute = numerical_cols[df[numerical_cols].isnull().median() < 0.30]
    
    # Apply KNN imputation only on the selected numerical columns
    df[cols_to_impute] = imputer.fit_transform(df[cols_to_impute])


In [20]:
# Create a dataframe to display the percentage of missingness for each feature year by year
missingness_df = pd.DataFrame()

for year, df in full_df.items():
    missingness_df[year] = df.isnull().mean() * 100

# Rearrange the index to display features (columns) and the percentage of missingness year by year
missingness_df.index.name = 'Feature'
missingness_df

2014       2015       2016  \
Feature                                                                       
Symbol                                       0.000000   0.000000   0.000000   
Revenue                                      0.000000   0.000000   0.000000   
Revenue Growth                               0.000000   0.000000   0.000000   
Cost of Revenue                              0.000000   0.000000   0.000000   
Gross Profit                                 0.000000   0.000000   0.000000   
R&D Expenses                                 0.000000   0.000000   0.000000   
SG&A Expense                                 0.000000   0.000000   0.000000   
Operating Expenses                           0.000000   0.000000   0.000000   
Operating Income                             0.000000   0.000000   0.000000   
Interest Expense                             0.000000   0.000000   0.000000   
Earnings before Tax                          0.000000   0.000000   0.000000   
Income Tax Expense                           0.000000   0.000000   0.000000   
Net Income - Non-Controlling int             0.000000   0.000000   0.000000   
Net Income - Discontinued ops                0.000000   0.000000   0.000000   
Net Income                                   0.000000   0.000000   0.000000   
Preferred Dividends                          0.000000   0.000000   0.000000   
Net Income Com                               0.000000   0.000000   0.000000   
EPS                                          0.000000   0.000000   0.000000   
EPS Diluted                                  0.000000   0.000000   0.000000   
Weighted Average Shs Out                     0.000000   0.000000   0.000000   
Weighted Average Shs Out (Dil)               0.000000   0.000000   0.000000   
Dividend per Share                           0.000000   0.000000   0.000000   
Gross Margin                                 0.000000   0.000000   0.000000   
EBITDA Margin                                0.000000   0.000000   0.000000   
EBIT Margin                                  0.000000   0.000000   0.000000   
Profit Margin                                0.000000   0.000000   0.000000   
Free Cash Flow margin                        0.000000   0.000000   0.000000   
EBITDA                                       0.000000   0.000000   0.000000   
EBIT                                         0.000000   0.000000   0.000000   
Consolidated Income                          0.000000   0.000000   0.000000   
Earnings Before Tax Margin                   0.000000   0.000000   0.000000   
Net Profit Margin                            0.000000   0.000000   0.000000   
Cash and cash equivalents                    0.000000   0.000000   0.000000   
Short-term investments                       0.000000   0.000000   0.000000   
Cash and short-term investments              0.000000   0.000000   0.000000   
Receivables                                  0.000000   0.000000   0.000000   
Inventories                                  0.000000   0.000000   0.000000   
Total current assets                         0.000000   0.000000   0.000000   
Property, Plant & Equipment Net              0.000000   0.000000   0.000000   
Goodwill and Intangible Assets               0.000000   0.000000   0.000000   
Long-term investments                        0.000000   0.000000   0.000000   
Tax assets                                   0.000000   0.000000   0.000000   
Total non-current assets                     0.000000   0.000000   0.000000   
Total assets                                 0.000000   0.000000   0.000000   
Payables                                     0.000000   0.000000   0.000000   
Short-term debt                              0.000000   0.000000   0.000000   
Total current liabilities                    0.000000   0.000000   0.000000   
Long-term debt                               0.000000   0.000000   0.000000   
Total debt                                   0.000000   0.000000   0.000000   
Deferred revenue    

### Dropping Features Due to Significant Missing Values

We are dropping the following features from our dataset as they contain a significant amount of missing data:

- **Total non-current assets**  
- **Return on Tangible Assets**
- **5Y Net Income Growth (per Share)**
- **5Y Operating CF Growth (per Share)**
- **5Y Shareholders Equity Growth (per Share)**  
- **5Y Dividend per Share Growth (per Share)**  
- **10Y Dividend per Share Growth (per Share)**
- **10Y Shareholders Equity Growth (per Share)**
- **10Y Net Income Growth (per Share)**
- **10Y Revenue Growth (per Share)**
- **10Y Operating CF Growth (per Share)**
- **Working Capital**
- **ROIC**
- **Current ratio**
- **Net Debt to EBITDA**
- **dividendPayoutRatio**
- **cashFlowCoverageRatios**
- **freeCashFlowOperatingCashFlowRatio**
- **cashFlowToDebtRatio**
- **Total non-current liabilities**
- **Net Debt**
- **Other Assets**
- **priceEarningsToGrowthRatio**
- **enterpriseValueMultiple**  
- **ebtperEBIT**  
- **niperEBT**  
- **effectiveTaxRate**
- **nIperEBT**  
- **eBTperEBIT**
- **operatingCycle**
- **cashConversionCycle**
- **priceCashFlowRatio**
- **returnOnAssets**
- **returnOnCapitalEmployed**
- **shortTermCoverageRatios**
- **5Y Revenue Growth (per Share)**

While some of these features are critical for assessing a company’s financial health and operational performance, such as **Working Capital**, **ROIC**, and **Current Ratio**, their high levels of missing data make them unsuitable for inclusion in the analysis. This may result in a less accurate model since these variables could provide important insights into the financial standing of the companies we are analyzing. 


In [21]:
# List of columns to drop
columns_to_drop = [
    'Total non-current assets', 
    'Return on Tangible Assets',
    '5Y Net Income Growth (per Share)',
    '5Y Operating CF Growth (per Share)',
    '5Y Shareholders Equity Growth (per Share)',  
    '5Y Dividend per Share Growth (per Share)',  
    '10Y Dividend per Share Growth (per Share)',
    '10Y Shareholders Equity Growth (per Share)',
    '10Y Net Income Growth (per Share)',
    '10Y Revenue Growth (per Share)',
    '10Y Operating CF Growth (per Share)',
    'Working Capital',
    'ROIC',
    'Current ratio',
    'Net Debt to EBITDA',
    'dividendPayoutRatio',
    'cashFlowCoverageRatios',
    'freeCashFlowOperatingCashFlowRatio',
    'cashFlowToDebtRatio',
    'Total non-current liabilities',
    'Net Debt',
    'Other Assets',
    'priceEarningsToGrowthRatio',
    'enterpriseValueMultiple',  
    'ebtperEBIT',  
    'niperEBT',  
    'effectiveTaxRate',
    'nIperEBT',  
    'eBTperEBIT',
    'operatingCycle',
    'cashConversionCycle',
    'priceCashFlowRatio',
    'returnOnAssets',
    'returnOnCapitalEmployed',
    'shortTermCoverageRatios',
    '5Y Revenue Growth (per Share)'
]

# Drop the columns from each dataset
for year, df in full_df.items():
    df.drop(columns=columns_to_drop, inplace=True, errors='ignore')

# Check missingness again after dropping the columns
missingness_after_drop = pd.DataFrame()

for year, df in full_df.items():
    missingness_after_drop[year] = df.isnull().mean() * 100

missingness_after_drop

2014  2015  2016  2017  2018
Symbol                                      0.0   0.0   0.0   0.0   0.0
Revenue                                     0.0   0.0   0.0   0.0   0.0
Revenue Growth                              0.0   0.0   0.0   0.0   0.0
Cost of Revenue                             0.0   0.0   0.0   0.0   0.0
Gross Profit                                0.0   0.0   0.0   0.0   0.0
R&D Expenses                                0.0   0.0   0.0   0.0   0.0
SG&A Expense                                0.0   0.0   0.0   0.0   0.0
Operating Expenses                          0.0   0.0   0.0   0.0   0.0
Operating Income                            0.0   0.0   0.0   0.0   0.0
Interest Expense                            0.0   0.0   0.0   0.0   0.0
Earnings before Tax                         0.0   0.0   0.0   0.0   0.0
Income Tax Expense                          0.0   0.0   0.0   0.0   0.0
Net Income - Non-Controlling int            0.0   0.0   0.0   0.0   0.0
Net Income - Discontinued ops               0.0   0.0   0.0   0.0   0.0
Net Income                                  0.0   0.0   0.0   0.0   0.0
Preferred Dividends                         0.0   0.0   0.0   0.0   0.0
Net Income Com                              0.0   0.0   0.0   0.0   0.0
EPS                                         0.0   0.0   0.0   0.0   0.0
EPS Diluted                                 0.0   0.0   0.0   0.0   0.0
Weighted Average Shs Out                    0.0   0.0   0.0   0.0   0.0
Weighted Average Shs Out (Dil)              0.0   0.0   0.0   0.0   0.0
Dividend per Share                          0.0   0.0   0.0   0.0   0.0
Gross Margin                                0.0   0.0   0.0   0.0   0.0
EBITDA Margin                               0.0   0.0   0.0   0.0   0.0
EBIT Margin                                 0.0   0.0   0.0   0.0   0.0
Profit Margin                               0.0   0.0   0.0   0.0   0.0
Free Cash Flow margin                       0.0   0.0   0.0   0.0   0.0
EBITDA                                      0.0   0.0   0.0   0.0   0.0
EBIT                                        0.0   0.0   0.0   0.0   0.0
Consolidated Income                         0.0   0.0   0.0   0.0   0.0
Earnings Before Tax Margin                  0.0   0.0   0.0   0.0   0.0
Net Profit Margin                           0.0   0.0   0.0   0.0   0.0
Cash and cash equivalents                   0.0   0.0   0.0   0.0   0.0
Short-term investments                      0.0   0.0   0.0   0.0   0.0
Cash and short-term investments             0.0   0.0   0.0   0.0   0.0
Receivables                                 0.0   0.0   0.0   0.0   0.0
Inventories                                 0.0   0.0   0.0   0.0   0.0
Total current assets                        0.0   0.0   0.0   0.0   0.0
Property, Plant & Equipment Net             0.0   0.0   0.0   0.0   0.0
Goodwill and Intangible Assets              0.0   0.0   0.0   0.0   0.0
Long-term investments                       0.0   0.0   0.0   0.0   0.0
Tax assets                                  0.0   0.0   0.0   0.0   0.0
Total assets                                0.0   0.0   0.0   0.0   0.0
Payables                                    0.0   0.0   0.0   0.0   0.0
Short-term debt                             0.0   0.0   0.0   0.0   0.0
Total current liabilities                   0.0   0.0   0.0   0.0   0.0
Long-term debt                              0.0   0.0   0.0   0.0   0.0
Total debt                                  0.0   0.0   0.0   0.0   0.0
Deferred revenue                            0.0   0.0   0.0   0.0   0.0
Tax Liabilities                             0.0   0.0   0.0   0.0   0.0
Deposit Liabilities                         0.0   0.0   0.0   0.0   0.0
Total liabilities                           0.0   0.0   0.0   0.0   0.0
Other comprehensive income                  0.0   0.0   0.0   0.0   0.0
Retained earnings (deficit)                 0.0   0.0   0.0   0.0   0.0
Total shareholders equity                   0.0   0.0   0.0   0.0   0.0
Investments

In [22]:
df.shape

(4392, 190)

##### Now we have 190 features we will work with those later on our model.

##### Let's save the cleaned dataset.

In [23]:
# Save the cleaned datasets to the 'data/cleaned_datasets/' directory
for year, df in full_df.items():
    file_name = f"data/cleaned_data/{year}_Financial_Data_Cleaned.csv"
    df.to_csv(file_name, index=False)
    print(f"{year}_Financial_Data_Cleaned.csv saved successfully")


2014_Financial_Data_Cleaned.csv saved successfully
2015_Financial_Data_Cleaned.csv saved successfully
2016_Financial_Data_Cleaned.csv saved successfully
2017_Financial_Data_Cleaned.csv saved successfully
2018_Financial_Data_Cleaned.csv saved successfully


## 5. Next Steps: Exploratory Data Analysis (EDA)

With the data cleaning process complete, the next step is to dive into **Exploratory Data Analysis (EDA)**. In this phase, we will explore the characteristics of our dataset in detail to uncover patterns, relationships, and insights that will inform our modeling approach.

The key steps in EDA will include:

1. **Descriptive Statistics**: Summarize the central tendency, dispersion, and shape of the dataset’s distribution. This will include metrics like mean, median, standard deviation, and more for key numerical features.
   
2. **Data Visualization**: Use various plots (histograms, box plots, scatter plots, heatmaps, etc.) to visualize the distribution of variables, relationships between features, and potential outliers.

3. **Correlation Analysis**: Explore the relationships between features, identifying highly correlated variables that may be relevant for model building.

4. **Outlier Detection**: Identify and analyze outliers that could affect the performance of the machine learning model. We will assess whether to handle these outliers by removing or transforming them.